src/static_array.cr

```diff
@@ -163,6 +163,16 @@ struct StaticArray(T, N)
     self
   end
 
+  # Like `map`, but the block gets passed both the element and its index.
+  def map_with_index!
+    i = 0
+    to_unsafe.map!(size) do |e|
+      i += 1
```

Comment:
Incrementing `i` after `yield` would allow to drop `i - 1` part which seems bit odd... just a thought :)